chapter/methodology.md

```diff
@@ -0,0 +1,1166 @@
+```
+\els
+
+
+
+## Illustrative example
+
+The example [@lst:nickel-complete-example] shows an illustrative high level configuration of a server.
+Using Nickel, this file would be used to define the schema of the configuration format of another program.
+Evaluation and validation is done in the context of Nickel, after which the evaluated structure is translated into a more common (but less expressive) format such as YAML or JSON.
+Here, the chema for a configuration of a Kubernetes-like [@kubernetes] tool is defined using contracts, making exemplary use of variables and functions.
+Specifically, it describes a way to provision named containers.
+The user is able to specify container images and opened ports, as well as define metadata of the deployment.
+The configuration is constrained by the `NobernetesConfig` contract.
+The contract in turn defines the required fields and field types.
+Notably, the fields `containers` and `replicas` are further constrained by individual contracts.
+The `Port` contract is a logical contract that ensures the value is in the range of valid port numbers.
+The example also shows different ways of declaring types (i.e. constraining record value types), string interpolation, as well as the usage of let bindings with standard types.
+Throughout this chapter, different sections about the NSL implementation will refer back to this example.
+
+```{.nickel #lst:nickel-complete-example caption="Nickel example with most features shown"}
+let Port | doc "A contract for a port number" =
+  contracts.from_predicate (fun value =>
+    builtins.is_num value &&
+    value % 1 == 0 &&
+    value >= 0 &&
+    value <= 65535) in
+
+let Container = {
+  image | Str,
+  ports | List Port,
+} in
+
+let NobernetesConfig = {
+  apiVersion | Str,
+  metadata.name | Str,
+  replicas | nums.PosNat
+           | doc "The number of replicas"
+           | default = 1,
+  containers | { _ : Container },
+
+} in
+
+let name_ = "myApp" in
+
+let metadata_ = {
+    name = name_,
+} in
+
+let webContainer = fun image => {
+  image = image,
+  ports = [ 80, 443 ],
+} in
+
+let image = "k8s.gcr.io/%{name_}" in
+
+{
+  apiVersion = "1.1.0",
+  metadata = metadata_,
+  replicas = 3,
+  containers = {
+    "main container" = webContainer image
+  }
+} | NobernetesConfig
+
+```
+
+
+## Linearization
+
+The focus of the NLS as presented in this work is to implement a foundational set of LSP features as described in [@sec:capability].
+In order to process these capabilities efficiently as per [@sec:performance], NLS needs to store more information than what is originally present in a Nickel AST (cf. [@sec:nickel-ast]), such as information about references these can be deduced from the AST lazily, it would require the repeated traversal of arbitrarily large tree with an associated cost to performance.
```

Comment:
sentence too long, split

Reply:
```suggestion
In order to process these capabilities efficiently as per [@sec:performance], NLS needs to store more information than what is originally present in a Nickel AST (cf. [@sec:nickel-ast]), such as information about references.
Although this can be deduced from the AST lazily, working with Nickel's tree representation is inefficient, as it is not optimized for random access and search operations.
```